core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -507,83 +507,92 @@ class KafkaApis(val requestChannel: RequestChannel,
+          if (!metadataCache.contains(topicPartition))
+            erroneous += topicPartition -> errorResponse(Errors.UNKNOWN_TOPIC_OR_PARTITION)
+          else
             interesting += (topicPartition -> data)
-          }
-        })
+        }
       } else {
-        fetchContext.foreachPartition((part, _) => {
-          erroneous += part -> new FetchResponse.PartitionData(Errors.TOPIC_AUTHORIZATION_FAILED,
-            FetchResponse.INVALID_HIGHWATERMARK, FetchResponse.INVALID_LAST_STABLE_OFFSET,
-            FetchResponse.INVALID_LOG_START_OFFSET, null, MemoryRecords.EMPTY)
-        })
+        fetchContext.foreachPartition { (part, _) =>
+          erroneous += part -> errorResponse(Errors.TOPIC_AUTHORIZATION_FAILED)
+        }
       }
     } else {
       // Regular Kafka consumers need READ permission on each partition they are fetching.
-      fetchContext.foreachPartition((topicPartition, data) => {
+      fetchContext.foreachPartition { (topicPartition, data) =>
         if (!authorize(request.session, Read, Resource(Topic, topicPartition.topic, LITERAL)))
-          erroneous += topicPartition -> new FetchResponse.PartitionData(Errors.TOPIC_AUTHORIZATION_FAILED,
-            FetchResponse.INVALID_HIGHWATERMARK, FetchResponse.INVALID_LAST_STABLE_OFFSET,
-            FetchResponse.INVALID_LOG_START_OFFSET, null, MemoryRecords.EMPTY)
+          erroneous += topicPartition -> errorResponse(Errors.TOPIC_AUTHORIZATION_FAILED)
         else if (!metadataCache.contains(topicPartition))
-          erroneous += topicPartition -> new FetchResponse.PartitionData(Errors.UNKNOWN_TOPIC_OR_PARTITION,
-            FetchResponse.INVALID_HIGHWATERMARK, FetchResponse.INVALID_LAST_STABLE_OFFSET,
-            FetchResponse.INVALID_LOG_START_OFFSET, null, MemoryRecords.EMPTY)
+          erroneous += topicPartition -> errorResponse(Errors.UNKNOWN_TOPIC_OR_PARTITION)
         else
           interesting += (topicPartition -> data)
-      })
+      }
+    }
+
+    def convertRecords(tp: TopicPartition, unconvertedRecords: Records, downConvertMagic: Option[Byte]): BaseRecords = {
+      downConvertMagic.map { magic =>
+        trace(s"Down converting records from partition $tp to message format version $magic for fetch request from $clientId")
+        // Because down-conversion is extremely memory intensive, we want to try and delay the down-conversion as much
+        // as possible. With KIP-283, we have the ability to lazily down-convert in a chunked manner. The lazy, chunked
+        // down-conversion always guarantees that at least one batch of messages is down-converted and sent out to the
+        // client.
+        new LazyDownConversionRecords(tp, unconvertedRecords, magic, fetchContext.getFetchOffset(tp).get, time)
+      }.getOrElse(unconvertedRecords)
     }
 
-    def convertRecords(tp: TopicPartition, unconvertedRecords: Records): BaseRecords = {
+    def convertFetchedData(tp: TopicPartition,
```

Comment:
Perhaps we should call this `maybeConvertFetchedData` since the conversion may or may not happen depending on whether it is needed and permitted.